src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -580,14 +626,260 @@ static int GetRawKdfCount(in AlgorithmIdentifierAsn algorithmIdentifier)
+                // So, any attribute that specifies it means it's a machine key.
+                if (bagAttributes is not null)
+                {
+                    foreach (AttributeAsn attr in bagAttributes)
+                    {
+                        if (attr.AttrType == Oids.MsPkcs12MachineKeySet)
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            static CngMachineKeyState CheckMachineKeyPermissions()
+            {
+#if NET || NETFRAMEWORK
+                CngKey? cngKey = null;
+
+                try
+                {
+#if NET
+                    Debug.Assert(OperatingSystem.IsWindows());
 #endif
 
+                    cngKey = CngKey.Create(
+                        CngAlgorithm.Rsa,
+                        Guid.NewGuid().ToString("B"),
+                        new CngKeyCreationParameters
+                        {
+                            Provider = CngProvider.MicrosoftSoftwareKeyStorageProvider,
+                            KeyCreationOptions = CngKeyCreationOptions.MachineKey,
+                        });
+
+                    return CngMachineKeyState.Permitted;
+                }
+                catch (CryptographicException)
+                {
+                    return CngMachineKeyState.Denied;
+                }
+                finally
+                {
+                    cngKey?.Dispose();
+                }
+#else
+                // The netstandard builds shouldn't ever execute on a relevant configuration,
+                // so it's not worth adding a reference to the Cng package for this method
+                // to compile.
+                return CngMachineKeyState.Permitted;
+#endif
+            }
+
+            static bool HasCapiCsp(AttributeAsn[]? bagAttributes, out bool isRsa)
+            {
+                isRsa = false;
+
+                if (bagAttributes is not null)
+                {
+                    foreach (AttributeAsn attr in bagAttributes)
+                    {
+                        if (attr.AttrType == Oids.MsPkcs12KeyProviderName && attr.AttrValues?.Length > 0)
+                        {
+                            return HasCapiValue(attr.AttrValues[0].Span, out isRsa);
+                        }
+                    }
+                }
+
+                // If there is no provider name at all, then PFXImportCertStore will use the CAPI
+                // MS_BASE_PROV provider, so it's a CAPI key.
+                return true;
+
+                static bool HasCapiValue(ReadOnlySpan<byte> encodedAttribute, out bool isRsa)
+                {
+                    // All of the MS_*_PROV_W values from wincrypt.h, except MS_SCARD_PROV(_W).
+                    const string MS_DEF_PROV = "Microsoft Base Cryptographic Provider v1.0";
```

Comment:
Do we usually use caps-casing for local `const`s?

Reply:
> 12. We use PascalCasing to name all our constant local variables and fields. The only exception is for interop code where the constant value should exactly match the name and value of the code you are calling via interop.

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md

That felt to me like it applied here.